_posts/2019-06-05-quarkus-and-web-ui-development-mode.adoc

```diff
@@ -0,0 +1,110 @@
+----
+$mvn package -Dui.dev -Dui.proxy quarkus:dev
+----
+The `-Dui` system property enables a Maven profile to build the web application. The output directory for the web application `webapp/dist/webapp` gets bundled into the Quarkus application. We will talk about `-Dui.proxy` later on.
+
+Now in a browser go to http://localhost:8080 and you will see our application running. It contains a few different pages. Go to the one called `Rest` which makes a REST call. We can change the return value of the https://github.com/kabir/blog-quarkus-ui-development/blob/master/src/main/java/org/kabir/quarkus/ui/SampleResource.java[`SampleServlet.hello()`] method. If you refresh the page you will see the changes made reflected automatically!
+
+With what we have so far if we were to make changes to the user interface in our web application, we would need to stop Quarkus and run the above command to build the web application again and restart quarkus. This is not what we want, so we take advantage of Angular's https://angular.io/guide/build#proxying-to-a-backend-server[proxy]. https://facebook.github.io/create-react-app/docs/proxying-api-requests-in-development[React] and https://cli.vuejs.org/config/#devserver-proxy[Vue] have something similar.
+
+In another terminal window go into the `webapp/` folder of the cloned project, and run:
+----
+$yarn proxy
+----
+This will start the Angular development server in proxy mode on port 4200. Now if you go to http://localhost:4200 you should see the same application as you saw on port 8080, but served by the Angular development server. If you make any changes to any of the Angular components set up in https://github.com/kabir/blog-quarkus-ui-development/blob/master/webapp/src/app/app.component.ts[`app.component.ts`] you will see your changes reflected.
+
+This is great!!! We can now be super-productive and make changes in both our front-end and our back-end, and see the changes reflected immediately without needing to rebuild, repackage and restart our application!
+
+==== Adjustments Summary
+
+The example project has already been set up with the needed adjustments to do this. These are the main things which need doing:
+
+. Make sure we handle the Angular routes in the UI rather than on the back-end. This is important for the bundled version (running on port 8080).
+. Enable the Angular proxy
+.. Signal to the back-end that the front-end is running in a proxy, i.e. not served by us. This is important for when the back-end needs to redirect to a page in the front-end.
+
+We will not go into details of how the application is set up, but will instead focus on the changes needed to make this work well.
+
+==== Handle Angular Routes
+We need to add a filter, https://github.com/kabir/blog-quarkus-ui-development/blob/master/src/main/java/org/kabir/quarkus/ui/AngularRouteFilter.java[AngularRouteFilter]. The essence of its `doFilter()` method is:
```

Comment:
This is minor but I prefer to replace the We need to sentences with just action sentences.
"We need to add" -> "Add"

Reply:
ok